module/apmpgx/tracer.go

```diff
@@ -0,0 +1,173 @@
+type tracer struct {
+	// cfg is the pgx.ConnConfig which used for setting metadata in spans such as host, port, etc.
+	cfg *pgx.ConnConfig
+
+	// logger used for writing data to log. If it's nil, then data won't be written to log, and only spans will be created.
+	logger pgx.Logger
+}
+
+// Instrument is getting pgx.ConnConfig and wrap logger into tracer.
+// It's safe to pass nil logger into pgx.ConnConfig, if so, then only spans will be created
+func Instrument(cfg *pgx.ConnConfig) {
+	cfg.Logger = &tracer{
+		cfg:    cfg,
+		logger: cfg.Logger,
+	}
+}
+
+// Log is getting type of SQL expression from msg and run suitable trace.
+// If logger in tracer struct isn't nil, than log will be
+// written to your logger that implements pgx.Logger interface.
+func (t *tracer) Log(ctx context.Context, level pgx.LogLevel, msg string, data map[string]interface{}) {
+	if t.logger != nil {
+		t.logger.Log(ctx, level, msg, data)
+	}
+
+	switch msg {
+	case "Query", "Exec":
+		t.QueryTrace(ctx, data)
+	case "CopyFrom":
+		t.CopyTrace(ctx, data)
+	case "SendBatch":
+		t.BatchTrace(ctx, data)
+	}
+}
+
+// QueryTrace traces query and creates spans for them.
+func (t *tracer) QueryTrace(ctx context.Context, data map[string]interface{}) {
+	statement, ok := data["sql"].(string)
+	if !ok {
+		apm.CaptureError(ctx,
+			fmt.Errorf("%w: expect string, got: %T",
+				ErrInvalidType,
+				data["sql"],
+			),
+		).Send()
+		return
+	}
+
+	span, ok := t.startSpan(ctx, apmsql.QuerySignature(statement), querySpanType, data)
+	if !ok {
+		return
+	}
+	defer span.End()
+}
+
+// CopyTrace traces copy queries and creates spans for them.
+func (t *tracer) CopyTrace(ctx context.Context, data map[string]interface{}) {
+	tableName, ok := data["tableName"].(pgx.Identifier)
+	if !ok {
+		return
+	}
+
+	span, ok := t.startSpan(ctx,
+		fmt.Sprintf("COPY TO %s", strings.Join(tableName, ", ")),
+		copySpanType,
+		data,
+	)
+	defer span.End()
+}
+
+// BatchTrace traces batch execution and creates spans for the whole batch.
+func (t *tracer) BatchTrace(ctx context.Context, data map[string]interface{}) {
+	span, ok := t.startSpan(ctx, "BATCH", batchSpanType, data)
+	if !ok {
+		return
+	}
+	defer span.End()
+
+	if batchLen, ok := data["batchLen"].(int); ok {
+		span.Context.SetLabel("batch.length", batchLen)
+	}
+}
+
+func (t *tracer) startSpan(ctx context.Context, spanName, spanType string, data map[string]interface{}) (*apm.Span, bool) {
+	stop := time.Now()
+
+	duration, ok := data["time"].(time.Duration)
+	if !ok {
+		apm.CaptureError(ctx, ErrUnsupportedPgxVersion).Send()
+		return nil, false
+	}
+
+	span, _ := apm.StartSpanOptions(ctx, spanName, spanType, apm.SpanOptions{
+		Start:    stop.Add(-duration),
+		ExitSpan: true,
+	})
+
+	if span.Dropped() {
+		span.End()
+		return nil, false
+	}
+
+	span.Duration = duration
+	span.Context.SetDatabase(apm.DatabaseSpanContext{
+		Instance:  t.cfg.Database,
+		Statement: spanName,
```

Comment:
Should we add another parameter to `startSpan` for passing in the statement? For QueryTrace, spanName will not be the full statement/query. This field is meant to capture the full query.

Reply:
I've added this parameter, but not sure that it's required for Copy and Batch trace, so provide empty string for them. We'll discuss this today with @EpicStep